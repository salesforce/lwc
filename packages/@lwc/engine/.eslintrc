{
    "env": {
        "browser": true
    },

    "globals": {
        "process": true
    },

    "rules": {
        "lwc-internal/no-production-assert": "error"
    },

    "overrides": [
        {
            "files": [
                "**/__tests__/**"
            ],

            "rules": {
                "lwc-internal/no-production-assert": "off"
            }
        },
        {
            "files": [ "src/polyfills/**"],
            "rules": {
                "no-extend-native": "off"
            }
        }
    ]
}
