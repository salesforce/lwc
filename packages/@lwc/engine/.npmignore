coverage/
scripts/
test/
src/
tsconfig.json
global.d.ts
yarn.lock
.*
