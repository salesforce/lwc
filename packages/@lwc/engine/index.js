// eslint-disable-next-line no-console
console.log('@lwc/engine cannot be imported via cjs');
