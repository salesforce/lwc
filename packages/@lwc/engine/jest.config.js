/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const path = require('path');
const BASE_CONFIG = require('../../../scripts/jest/base.config');

module.exports = {
    ...BASE_CONFIG,

    displayName: 'lwc-engine',

    roots: ['<rootDir>/src'],

    // Customize setup for the engine tests.
    setupFilesAfterEnv: [path.resolve(__dirname, 'scripts/jest/setup-test.js')],
    moduleNameMapper: {
        'test-utils': path.resolve(__dirname, 'scripts/jest/test-utils.js'),
    },

    // Ignore jest custom setup scripts from the code coverage.
    coveragePathIgnorePatterns: [
        '<rootDir>/scripts/',
        '<rootDir>/src/polyfills',
        '<rootDir>/src/framework/wc.ts',
    ],

    // Override global threshold for the package. As we increase the test coverage we should increase
    // the threshold as well.
    coverageThreshold: {
        global: {
            functions: 65,
            lines: 75,
            branches: 70,
        },
    },
};
