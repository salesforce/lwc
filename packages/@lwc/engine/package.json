{
    "name": "@lwc/engine",
    "version": "1.0.2",
    "description": "LWC Engine",
    "main": "dist/engine.cjs.js",
    "module": "dist/engine.js",
    "typings": "types/index.d.ts",
    "license": "MIT",
    "scripts": {
        "clean": "rm -rf dist/ types/",
        "test": "jest",
        "build": "tsc --emitDeclarationOnly && rollup --config scripts/rollup/rollup.config.js"
    },
    "files": [
        "dist/",
        "types/"
    ],
    "dependencies": {
        "@lwc/features": "1.0.2",
        "@lwc/shared": "1.0.2",
        "observable-membrane": "0.26.1"
    },
    "devDependencies": {
        "@lwc/template-compiler": "1.0.2"
    },
    "lwc": {
        "modules": [
            {
                "name": "lwc",
                "path": "dist/engine.js"
            }
        ]
    },
    "publishConfig": {
        "access": "public"
    }
}
