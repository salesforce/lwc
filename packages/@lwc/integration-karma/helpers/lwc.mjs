import path from 'node:path';
import { rollup } from 'rollup';
import lwcRollupPlugin from '@lwc/rollup-plugin';

import {
    DISABLE_SYNTHETIC_SHADOW_SUPPORT_IN_COMPILER,
    API_VERSION,
    DISABLE_STATIC_CONTENT_OPTIMIZATION,
} from './options.mjs';

// Cache reused between each compilation to speed up the compilation time.
let cache;

export default async (ctx) => {
    const input = ctx.path.slice(1); // strip leading / from URL path to get relative file path
    const suiteDir = path.dirname(input);

    // TODO [#3370]: remove experimental template expression flag
    const experimentalComplexExpressions = suiteDir.includes('template-expressions');

    const createRollupPlugin = (options) => {
        return lwcRollupPlugin({
            // Sourcemaps don't work with Istanbul coverage
            sourcemap: !process.env.COVERAGE,
            experimentalDynamicComponent: {
                loader: 'test-utils',
                strict: true,
            },
            enableDynamicComponents: true,
            enableLwcOn: true,
            experimentalComplexExpressions,
            enableStaticContentOptimization: !DISABLE_STATIC_CONTENT_OPTIMIZATION,
            disableSyntheticShadowSupport: DISABLE_SYNTHETIC_SHADOW_SUPPORT_IN_COMPILER,
            apiVersion: API_VERSION,
            ...options,
        });
    };

    const defaultRollupPlugin = createRollupPlugin();

    const customLwcRollupPlugin = {
        ...defaultRollupPlugin,
        transform(src, id) {
            let rollupPluginToUse;

            // Override the LWC Rollup plugin to specify different options based on file name patterns.
            // This allows us to alter the API version or other compiler props on a filename-only basis.
            const apiVersion = id.match(/useApiVersion(\d+)/)?.[1];
            const nativeOnly = /\.native-only\./.test(id);
            if (apiVersion) {
                rollupPluginToUse = createRollupPlugin({
                    apiVersion: parseInt(apiVersion, 10),
                });
            } else if (nativeOnly) {
                rollupPluginToUse = createRollupPlugin({ disableSyntheticShadowSupport: true });
            } else {
                rollupPluginToUse = defaultRollupPlugin;
            }
            return rollupPluginToUse.transform.call(this, src, id);
        },
    };

    const plugins = [customLwcRollupPlugin];

    const bundle = await rollup({
        input,
        plugins,
        cache,

        // Rollup should not attempt to resolve the engine and the test utils, Karma takes care of injecting it
        // globally in the page before running the tests.
        external: ['lwc', 'wire-service', 'test-utils', '@test/loader'],

        onwarn(warning, warn) {
            // Ignore warnings from our own Rollup plugin
            if (warning.plugin !== 'rollup-plugin-lwc-compiler') {
                warn(warning);
            }
        },
    });

    cache = bundle.cache;

    const { output } = await bundle.generate({
        format: 'esm',
        // TODO: Does web-test-runner use istanbul?
        // Sourcemaps don't work with Istanbul coverage
        sourcemap: process.env.COVERAGE ? false : 'inline',

        // The engine and the test-utils is injected as UMD. This mapping defines how those modules can be
        // referenced from the window object.
        globals: {
            lwc: 'LWC',
            'wire-service': 'WireService',
            'test-utils': 'TestUtils',
        },
    });

    const { code } = output[0];
    return code;
};
