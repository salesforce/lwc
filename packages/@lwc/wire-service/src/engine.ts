/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
// subtypes from lwc
export interface WireDef {
    params?: {
        [key: string]: string;
    };
    static?: {
        [key: string]: any;
    };
    adapter: any;
    method?: 1;
}
export interface ElementDef {
    // wire is optional on ElementDef but the lwc guarantees it before invoking wiring service hook
    wire: {
        [key: string]: WireDef;
    };
}
