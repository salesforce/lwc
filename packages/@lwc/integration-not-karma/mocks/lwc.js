// IMPORTANT: we must use @lwc/engine-dom instead of lwc in order to avoid circular imports
import { sanitizeAttribute as _sanitizeAttribute } from '@lwc/engine-dom';
import { fn } from '@vitest/spy';

export * from '@lwc/engine-dom';
export const sanitizeAttribute = fn(_sanitizeAttribute);
