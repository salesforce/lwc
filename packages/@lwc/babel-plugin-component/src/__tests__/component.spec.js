/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const pluginTest = require('./utils/test-transform').pluginTest(require('../index'));

describe('Element import', () => {
    pluginTest(
        'throws if using default import on lwc',
        `
        import engine from 'lwc';
    `,
        {
            error: {
                message: `Invalid import. "lwc" doesn't have default export.`,
                loc: {
                    line: 1,
                    column: 7,
                    length: 6,
                    start: 7,
                },
            },
        }
    );

    pluginTest(
        'throws if using namespace import on lwc',
        `
        import * as engine from 'lwc';
        export default class extends engine.LightningElement {}
    `,
        {
            error: {
                message: `Invalid import. Namespace imports are not allowed on "lwc", instead use named imports "import { LightningElement } from 'lwc'".`,
                loc: {
                    line: 1,
                    column: 7,
                    length: 11,
                    start: 7,
                },
            },
        }
    );

    pluginTest(
        'allow to remap the import to LightningElement',
        `
        import { LightningElement as Component } from 'lwc';

        export default class Test extends Component {}
    `,
        {
            output: {
                code: `
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";
                import { LightningElement as Component } from "lwc";

                class Test extends Component {}

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        'throws non-whitelisted lwc api is imported',
        `
        import { registerTemplate } from "lwc";
        import tmpl from './localTemplate.html';
        registerTemplate(tmpl);
    `,
        {
            error: {
                message: `Invalid import. "registerTemplate" is not part of the lwc api.`,
                loc: {
                    line: 1,
                    column: 0,
                    length: 104,
                    start: 0,
                },
            },
        }
    );

    pluginTest(
        'allows importing whitelisted apis from "lwc"',
        `
        import {
            api,
            track,
            wire,
            createElement,
            LightningElement,
            buildCustomElementConstructor,
            getComponentDef,
            getComponentConstructor,
            isComponentConstructor,
            readonly,
            register,
            unwrap,
        } from "lwc";
    `,
        {
            output: {
                code: `import { api, track, wire, createElement, LightningElement, buildCustomElementConstructor, getComponentDef, getComponentConstructor, isComponentConstructor, readonly, register, unwrap } from "lwc";`,
            },
        }
    );
});

describe('render method', () => {
    pluginTest(
        'inject render method',
        `
        import { LightningElement } from "lwc";
        export default class Test extends LightningElement {}
    `,
        {
            output: {
                code: `
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";
                import { LightningElement } from "lwc";

                class Test extends LightningElement {}

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        'does not insert render method when extending from legacy "engine" Element',
        `
        import { Element } from "engine";
        export default class Test extends Element {}
    `,
        {
            output: {
                code: `
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";
                import { Element } from "engine";

                class Test extends Element {}

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        `keep the render method if present`,
        `
        import { LightningElement } from "lwc";
        export default class Test extends LightningElement {
            render() {}
        }
    `,
        {
            output: {
                code: `
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";
                import { LightningElement } from "lwc";

                class Test extends LightningElement {
                  render() {}
                }

                export default _registerComponent(Test, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );

    pluginTest(
        'only inject render in the exported class',
        `
        import { LightningElement } from 'lwc';

        class Test1 extends LightningElement {}

        export default class Test2 extends LightningElement {}
    `,
        {
            output: {
                code: `
                import _tmpl from "./test.html";
                import { registerComponent as _registerComponent } from "lwc";
                import { LightningElement } from "lwc";

                class Test1 extends LightningElement {}

                class Test2 extends LightningElement {}

                export default _registerComponent(Test2, {
                  tmpl: _tmpl
                });
                `,
            },
        }
    );
});
