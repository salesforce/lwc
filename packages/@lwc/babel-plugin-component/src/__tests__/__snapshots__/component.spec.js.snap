// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element import throws if using default import on lwc 1`] = `
Object {
  "column": 7,
  "length": 6,
  "line": 1,
  "start": 7,
}
`;

exports[`Element import throws if using namespace import on lwc 1`] = `
Object {
  "column": 7,
  "length": 11,
  "line": 1,
  "start": 7,
}
`;

exports[`Element import throws non-whitelisted lwc api is imported 1`] = `
Object {
  "column": 0,
  "length": -1,
  "line": 1,
  "start": 0,
}
`;
